src/http_client_async.zig

```diff
@@ -772,6 +864,184 @@ pub const HTTPThread = struct {
 const log = Output.scoped(.fetch, false);
 
 var temp_hostname: [8096]u8 = undefined;
+
+const INET6_ADDRSTRLEN = if (bun.Environment.isWindows) 65 else 46;
+
+fn canonicalizeIP(addr_str: []const u8, outIP: *[INET6_ADDRSTRLEN + 1]u8) ![]const u8 {
+    if (addr_str.len >= INET6_ADDRSTRLEN) {
+        return error.Invalid;
+    }
+    var ip_std_text: [INET6_ADDRSTRLEN + 1]u8 = undefined;
+    // we need a null terminated string as input
+    bun.copy(u8, outIP, addr_str);
+    outIP[addr_str.len] = 0;
+
+    var af: c_int = std.os.AF.INET;
+    // get the standard text representation of the IP
+    if (c_ares.ares_inet_pton(af, outIP, &ip_std_text) != 1) {
+        af = std.os.AF.INET6;
+        if (c_ares.ares_inet_pton(af, outIP, &ip_std_text) != 1) {
+            return error.Invalid;
+        }
+    }
+    // ip_addr will contain the null-terminated string of the cannonicalized IP
+    if (c_ares.ares_inet_ntop(af, &ip_std_text, outIP, outIP.len) == null) {
+        return error.Invalid;
+    }
+    // use the null-terminated size to return the string
+    const size = bun.len(bun.cast([*:0]u8, outIP));
+    return outIP[0..size];
+}
+
+fn ipv4ToString(ipv4Bytes: *[4]u8, outIP: *[INET6_ADDRSTRLEN + 1]u8) ![]const u8 {
+    var ipv4String: [16]u8 = undefined;
+    return try canonicalizeIP(try std.fmt.bufPrint(&ipv4String, "{}.{}.{}.{}", .{ ipv4Bytes[0], ipv4Bytes[1], ipv4Bytes[2], ipv4Bytes[3] }), outIP);
+}
+
+const hexChars = "0123456789ABCDEF";
+fn ipv6ToString(ipv6Bytes: *[16]u8, outIP: *[INET6_ADDRSTRLEN + 1]u8) ![]const u8 {
```

Comment:
i'm a little worried about edgecases that lead to security issues in this code

can we rely more on c-ares for this?

Reply:
sure need to check c-ares manual

Reply:
fixed